File: scripts/mcp_stdio_entrypoint.py
import os
import sys
import json
import httpx
from mcp.server.fastmcp import FastMCP

# --- MCPClient Class for Backend Integration ---
class MCPClient:
    def __init__(self, base_url: str, api_key: str):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.client = httpx.AsyncClient(headers={"X-API-Key": api_key})

    async def get_memory_context(self, query: str, max_items: int = 10) -> dict:
        url = f"{self.base_url}/api/v1/memory/context"
        try:
            response = await self.client.post(url, json={"query": query, "max_items": max_items}, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def search_knowledge_base(self, query: str) -> dict:
        # This endpoint does not exist yet, so we will simulate it by calling the existing one.
        # In a real implementation, you would have a dedicated endpoint for this.
        return await self.get_memory_context(query=query)

    async def search_conversations(self, query: str) -> dict:
        # This endpoint does not exist yet, so we will simulate it by calling the existing one.
        return await self.get_memory_context(query=query)

    async def add_documents(self, documents: list) -> dict:
        url = f"{self.base_url}/api/v1/knowledge/documents"
        try:
            response = await self.client.post(url, json={"documents": documents}, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def get_current_conversation(self) -> dict:
        url = f"{self.base_url}/api/v1/conversation/current"
        try:
            response = await self.client.get(url, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def end_conversation(self) -> dict:
        url = f"{self.base_url}/api/v1/conversation/end"
        try:
            response = await self.client.post(url, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def get_memory_stats(self) -> dict:
        url = f"{self.base_url}/api/v1/memory/stats"
        try:
            response = await self.client.get(url, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def list_embedding_models(self) -> dict:
        url = f"{self.base_url}/api/v1/embeddings/models"
        try:
            response = await self.client.get(url, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def switch_embedding_model(self, model_name: str) -> dict:
        url = f"{self.base_url}/api/v1/embeddings/switch"
        try:
            response = await self.client.post(url, json={"model_name": model_name}, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPError as e:
            raise Exception(f"HTTP Request failed: {e}")

    async def close(self):
        await self.client.aclose()

# --- Initialize FastMCP Server ---
api_key = os.getenv("MCP_API_KEY")
if not api_key:
    raise ValueError("MCP_API_KEY environment variable not set")

base_url = "https://ai.avengergear.com"
client = MCPClient(base_url=base_url, api_key=api_key)
mcp = FastMCP("mojo-assistant")

# --- Define Tools ---
@mcp.tool()
async def get_memory_context(query: str, max_items: int = 10) -> str:
    """Search all memory tiers for relevant context."""
    result = await client.get_memory_context(query=query, max_items=max_items)
    return json.dumps(result)

@mcp.tool()
async def search_knowledge_base(query: str) -> str:
    """Search for information exclusively within the knowledge base."""
    result = await client.search_knowledge_base(query=query)
    return json.dumps(result)

@mcp.tool()
async def search_conversations(query: str) -> str:
    """Search for past conversations in the archival memory."""
    result = await client.search_conversations(query=query)
    return json.dumps(result)

@mcp.tool()
async def add_documents(documents: list) -> str:
    """Add documents to the knowledge base."""
    result = await client.add_documents(documents=documents)
    return json.dumps(result)

@mcp.tool()
async def get_current_conversation() -> str:
    """Get the messages from the current working memory."""
    result = await client.get_current_conversation()
    return json.dumps(result)

@mcp.tool()
async def end_conversation() -> str:
    """End the current conversation and archive it to memory."""
    result = await client.end_conversation()
    return json.dumps(result)

@mcp.tool()
async def get_memory_stats() -> str:
    """Get comprehensive statistics about the memory system."""
    result = await client.get_memory_stats()
    return json.dumps(result)

@mcp.tool()
async def list_embedding_models() -> str:
    """List the available embedding models."""
    result = await client.list_embedding_models()
    return json.dumps(result)

@mcp.tool()
async def switch_embedding_model(model_name: str) -> str:
    """Switch to a different embedding model."""
    result = await client.switch_embedding_model(model_name=model_name)
    return json.dumps(result)

# --- Run the Server ---
if __name__ == "__main__":
    try:
        mcp.run(transport="stdio")
    except Exception as e:
        print(f"Server error: {e}", file=sys.stderr)
        sys.stderr.flush()
        raise
    finally:
        import anyio
        anyio.run(client.close)

File: mcp_ts_client/src/mcp_stdio_entrypoint.ts
import axios, { AxiosInstance } from 'axios';
import * as fs from 'fs';
import * as readline from 'readline';

// --- MCPClient Class for Backend Integration ---
class MCPClient {
    private client: AxiosInstance;

    constructor(private baseUrl: string, private apiKey: string) {
        this.baseUrl = baseUrl.replace(/\$/, '');
        this.client = axios.create({
            baseURL: this.baseUrl,
            headers: { 'X-API-Key': this.apiKey },
            timeout: 30000,
        });
    }

    private async makeRequest(method: 'get' | 'post', endpoint: string, data?: any): Promise<any> {
        try {
            const response = await this.client[method](endpoint, data);
            return response.data;
        } catch (error) {
            if (axios.isAxiosError(error)) {
                throw new Error(`HTTP Request failed: ${error.message}`);
            }
            throw error;
        }
    }

    async getMemoryContext(query: string, maxItems: number = 10): Promise<any> {
        return this.makeRequest('post', '/api/v1/memory/context', { query, max_items: maxItems });
    }

    async searchKnowledgeBase(query: string): Promise<any> {
        return this.getMemoryContext(query); // Simulate
    }

    async searchConversations(query: string): Promise<any> {
        return this.getMemoryContext(query); // Simulate
    }

    async addDocuments(documents: any[]): Promise<any> {
        return this.makeRequest('post', '/api/v1/knowledge/documents', { documents });
    }

    async getCurrentConversation(): Promise<any> {
        return this.makeRequest('get', '/api/v1/conversation/current');
    }

    async endConversation(): Promise<any> {
        return this.makeRequest('post', '/api/v1/conversation/end');
    }

    async getMemoryStats(): Promise<any> {
        return this.makeRequest('get', '/api/v1/memory/stats');
    }

    async listEmbeddingModels(): Promise<any> {
        return this.makeRequest('get', '/api/v1/embeddings/models');
    }

    async switchEmbeddingModel(model_name: string): Promise<any> {
        return this.makeRequest('post', '/api/v1/embeddings/switch', { model_name });
    }
}

// --- Main Stdio Handling Logic ---
async function main() {
    const apiKey = process.env.MCP_API_KEY;
    const baseUrl = 'https://ai.avengergear.com';

    if (!apiKey) {
        console.error(JSON.stringify({ error: 'MCP_API_KEY environment variable not set' }));
        process.exit(1);
    }

    const client = new MCPClient(baseUrl, apiKey);

    let apiDescription: any;
    try {
        const configPath = process.env.MCP_CONFIG_PATH || `${process.env.HOME}/.config/mcp/mojo-assistant`;
        apiDescription = JSON.parse(fs.readFileSync(`${configPath}/mcp_api_description.json`, 'utf-8'));
    } catch (error) {
        console.error(JSON.stringify({ error: `mcp_api_description.json not found: ${error}` }));
        process.exit(1);
    }

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: false,
    });

    for await (const line of rl) {
        try {
            const request = JSON.parse(line);
            let response: any;
            const requestId = request.id;

            switch (request.method) {
                case 'initialize':
                    response = { jsonrpc: '2.0', id: requestId, result: { capabilities: { tool_provider: true } } };
                    break;

                case 'ListTools':
                    response = { jsonrpc: '2.0', id: requestId, result: { tools: apiDescription } };
                    break;

                case 'CallTool':
                    const { name, arguments: args } = request.params;
                    try {
                        let result;
                        switch (name) {
                            case 'get_memory_context': result = await client.getMemoryContext(args.query, args.max_items); break;
                            case 'search_knowledge_base': result = await client.searchKnowledgeBase(args.query); break;
                            case 'search_conversations': result = await client.searchConversations(args.query); break;
                            case 'add_documents': result = await client.addDocuments(args.documents); break;
                            case 'get_current_conversation': result = await client.getCurrentConversation(); break;
                            case 'end_conversation': result = await client.endConversation(); break;
                            case 'get_memory_stats': result = await client.getMemoryStats(); break;
                            case 'list_embedding_models': result = await client.listEmbeddingModels(); break;
                            case 'switch_embedding_model': result = await client.switchEmbeddingModel(args.model_name); break;
                            default: throw new Error(`Unknown tool: ${name}`);
                        }
                        response = { jsonrpc: '2.0', id: requestId, result: { content: [{ type: 'text', text: JSON.stringify(result) }] } };
                    } catch (e) {
                        response = { jsonrpc: '2.0', id: requestId, error: { code: -32000, message: (e as Error).message } };
                    }
                    break;
            }

            if (response) {
                process.stdout.write(JSON.stringify(response) + '\n');
            }
        } catch (e) {
            console.error(JSON.stringify({ error: `Unhandled error in main loop: ${e}` }));
        }
    }
}

main().catch(err => {
    console.error(JSON.stringify({ error: `Fatal error: ${err}` }));
    process.exit(1);
});
